### DISCUSSION TITLE
Implementation of BepInExPack

### DISCUSSION THREAD

**Discussion by madhub0815** [lloesche/valheim-server-docker#279] [Ideas]
Hello 

Are there any plans to implement BepInExPack similar as the Valheim+ mod?
That would be a great way for easy mod module loading. 

[Mod Link](https://valheim.thunderstore.io/package/denikson/BepInExPack_Valheim/)

**Comment by lloesche** [comment]
Just looked into it. There doesn't seem to be a good API available like there is for Github. I toyed around with their experimental API but all the download URLs just lead to a website not the actual ZIP archive.

E.g.
```
lukas@MACROMI:~/repo/valheim-server-docker$ curl -sX GET -H "accept: application/json" "https://thunderstore.io/api/experimental/package/denikson/BepInExPack_Valheim/" | jq .
{
  "namespace": "denikson",
  "name": "BepInExPack_Valheim",
  "full_name": "denikson-BepInExPack_Valheim",
  "owner": "denikson",
  "package_url": "https://thunderstore.io/package/denikson/BepInExPack_Valheim/",
  "date_created": "2021-02-14T18:07:34.498403Z",
  "date_updated": "2021-03-26T21:26:57.122167Z",
  "rating_score": 660,
  "is_pinned": true,
  "is_deprecated": false,
  "total_downloads": 17778024,
  "latest": {
    "namespace": "denikson",
    "name": "BepInExPack_Valheim",
    "version_number": "5.4.901",
    "full_name": "denikson-BepInExPack_Valheim-5.4.901",
    "description": "BepInEx pack for Valheim. Preconfigured and includes unstripped Unity DLLs.",
    "icon": "https://cdn.thunderstore.io/live/repository/icons/denikson-BepInExPack_Valheim-5.4.901.png",
    "dependencies": [],
    "download_url": "https://thunderstore.io/package/download/denikson/BepInExPack_Valheim/5.4.901/",
    "downloads": 15989,
    "date_created": "2021-03-26T21:26:56.689141Z",
    "website_url": "https://github.com/BepInEx/BepInEx",
    "is_active": true
  },
  "community_listings": [
    {
      "has_nsfw_content": false,
      "categories": [],
      "community": "valheim"
    }
  ]
}
lukas@MACROMI:~/repo/valheim-server-docker$ curl -sX GET -H "accept: application/json" "https://thunderstore.io/api/experimental/package/denikson/BepInExPack_Valheim/5.4.901/" | jq .
{
  "namespace": "denikson",
  "name": "BepInExPack_Valheim",
  "version_number": "5.4.901",
  "full_name": "denikson-BepInExPack_Valheim-5.4.901",
  "description": "BepInEx pack for Valheim. Preconfigured and includes unstripped Unity DLLs.",
  "icon": "https://cdn.thunderstore.io/live/repository/icons/denikson-BepInExPack_Valheim-5.4.901.png",
  "dependencies": [],
  "download_url": "https://thunderstore.io/package/download/denikson/BepInExPack_Valheim/5.4.901/",
  "downloads": 15983,
  "date_created": "2021-03-26T21:26:56.689141Z",
  "website_url": "https://github.com/BepInEx/BepInEx",
  "is_active": true
}
```
There's no direct link from where we could fetch the ZIP and extract it inside the container like we do for V+.

Guess their intention is for this to be shown in some download/mod manager and then people loading their website pre-download so they get the ad revenue. No obvious way of reliably automating it.

They are storing the files on e.g. `https://cdn.thunderstore.io/live/repository/packages/denikson-BepInExPack_Valheim-5.4.901.zip` which I suppose would be easy enough to construct ourselves just by taking the icon URL and replacing `icons` with `packages` and `png` with `zip`. But I have no intention of working around their API or doing something they didn't want users to do.

**Reply by lloesche** [reply]
I asked the Thunderstore authors if it's alright with them for us to pull the mod directly.
https://github.com/thunderstore-io/Thunderstore/issues/293

**Reply by lloesche** [reply]
So turns out that those download URLs will redirect to the correct ZIP archive and it's fine to fetch them directly. I'll create an issue to track the implementation.

**Reply by madhub0815** [reply]
Wow awesome I thank you for your great work. 😉👍

**Comment by gdragon-git** [comment]
One plugin that seems like a great potential pair is Valheim Online. ( https://www.nexusmods.com/valheim/mods/626 )
It's just one DLL file, but apart from having to figure out which hook to connect to do to a file copy of the DLL to the right location, there's an internal directory that would have to be exposed to allow for character imports or editing. 

More info on friction points in this post: https://forums.nexusmods.com/index.php?/topic/9793768-valheim-online-server-side-characters/page-6#entry92481123  

1. Users have to set up some kind of script to move in the ValheimOnline.dll after the right PRE or POST hook ( https://github.com/l...ker#event-hooks ). Not sure which hook, but I imagine it would have to copy from the right Docker Host OS folder location the dll and move it into {install}/data/plus/BepInEx/plugins (if permissions allow).
2. LLoesche would have to expose /.config/unity3d/IronGate/Valheim/character_vault/ through to the Docker host OS somehow, to allow for character imports.

**Comment by madhub0815** [comment]
> There's no direct link from where we could fetch the ZIP and extract it inside the container like we do for V+.
> 
> Guess their intention is for this to be shown in some download/mod manager and then people loading their website pre-download so they get the ad revenue. No obvious way of reliably automating it.

OK I understand.
Maybe https://github.com/BepInEx/BepInEx can be used for pull the zip file?

**Reply by lloesche** [reply]
The two are not the same. BepInEx and Valheim BepInExPack are two separate things. The later is BepInEx prepared for use with Valheim. The former is just the generic Unity Mod tool which is missing some libraries Valheim requires. That's why the BepInExPack author build that bundle in the first place.
